File: parsing/parser.py
from util.error import LanguageError
from .expressions.expression import Expression
from .node import Node
from lexing.token import Token

class ParserError(LanguageError):
    pass

class Parser:
    @property
    def expression(self):
        return Expression

    def __init__(self, tokens: list[Token]):
        self.tokens = tokens
        self.i = 0

    def next(self) -> Token:
        return self.tokens[self.i]

    def take(self) -> Token:
        token = self.next()
        self.i += 1
        return token

    def expecting_has(self, *args: str):
        if self.next().has(*args):
            return self.take()

        raise ParserError(f"Expecting {args}")

    def expecting_of(self, *args: str):
        if self.next().of(*args):
            return self.take()

        raise ParserError(f"Expecting {args}")

    def make_tree(self) -> Node:
        node = Expression.construct(self)

        if not self.next().has("EOF"):
            raise ParserError(f"Unexpected token {self.next()}")

        return node

File: app/src/main/kotlin/sea/stages/Parser.kt
package sea.stages

import kotlin.math.min
import sea.Faults
import sea.grammar.Token
import sea.grammar.FileStat

data class Parser(val faults: Faults, private val tokens: List<Token>) {
    init { faults.stageVerb = "Parsing" }

    private var i = 0
    val tree = FileStat()

    override fun toString() = "$tree"
    fun makeTree() = tree.makeTree(this)

    fun ahead(n: Int) = tokens[min(i + n, tokens.size - 1)]
    val next get() = ahead(0)

    fun take(): Token {
        val token = next
        i += 1
        return token
    }

    fun expectingOf(vararg types: Token.Type): Token {
        if(next.of(*types)) return take()

        val names = (types.map{it.name}).joinToString(", ", "[", "]")
        throw faults.fail(take(), "Expecting type of $names")
    }

    fun expectingHas(vararg values: String): Token {
        if(next.has(*values)) return take()

        val vals = values.joinToString(",", "[", "]")
        throw faults.fail(take(),
            "Expecting value of $vals".replace("\n", "\\n"))
    }

    fun expectingOf(c: Collection<Token.Type>) = expectingOf(*c.toTypedArray())
    fun expectingHas(c: Collection<String>) = expectingHas(*c.toTypedArray())
    fun skipNewline() { if(next.has("\n")) take() }
}
